snuba/query/snql/joins.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+from typing import MutableMapping, NamedTuple, Optional, Sequence, Union
+
+from snuba.datasets.entities import EntityKey
+from snuba.datasets.entities.factory import get_entity
+from snuba.query.data_source.join import (
+    IndividualNode,
+    JoinClause,
+    JoinCondition,
+    JoinConditionExpression,
+    JoinRelationship,
+)
+from snuba.query.data_source.simple import Entity as QueryEntity
+from snuba.query.parser.exceptions import ParsingException
+
+
+class EntityTuple(NamedTuple):
+    alias: str
+    name: str
+    sample_rate: Optional[float] = None
+
+
+class RelationshipTuple(NamedTuple):
+    lhs: EntityTuple
+    relationship: str
+    rhs: EntityTuple
+    data: JoinRelationship
+
+
+class Leaf(object):
+    def __init__(
+        self,
+        entity: EntityKey,
+        entity_data: EntityTuple,
+        relationship: Optional[JoinRelationship] = None,
+    ) -> None:
+        self.entity = entity
+        self.entity_data = entity_data
+        self.relationship = relationship
+        self.child: Optional[Leaf] = None
```

Comment:
If a node can have only one leaf, how are we expressing a graph where a root entity (events) is connected with two children (groupedmessage and groupassignee) ?

Reply:
It's a linked list of the form `events -> groupedmessage -> groupassignee`. The children store the join conditions so the two leafs will know they are joining events.